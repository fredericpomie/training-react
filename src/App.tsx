import React, {useReducer, useState} from "react";
import "./App.css";
import Brand from "./Brand";
import useBrands from "./useBrands";

export const BoldContext = React.createContext(false);
BoldContext.displayName = 'DisplayContext';

interface AppState {
  query: string;
  page: number;
}

type Action =
  | {type: 'PAGE_UP'}
  | {type: 'PAGE_DOWN'}
  | {type: 'QUERY_UPDATED', payload: string};

function paginationReducer(state: AppState, action: Action) {
  const draft = {...state};

  switch (action.type) {
    case "PAGE_UP":
      draft.page =  draft.page + 1;
      break;
    case "PAGE_DOWN":
      draft.page =  draft.page - 1;
      break;
    case "QUERY_UPDATED":
      draft.page =  1;
      draft.query = action.payload
      break;
    default:
      throw new Error();
  }

  return draft;
}

function App() {
  const [bold, setBold] = useState(false);
  const [{page, query}, dispatch] = useReducer(paginationReducer, {page: 1, query: ""});
  const brands = useBrands(page, query);

  return (
    <BoldContext.Provider value={bold}>
      <div className="App container">
        <div className="form-group">
          <label>Search</label>
          <input
            type="checkbox"
            onChange={({target}) => setBold(target.checked)}
            checked={bold}
          />
          <input
            type="search"
            className="form-control"
            value={query}
            onChange={({ target }) => {
              dispatch({type: 'QUERY_UPDATED', payload: target.value});
            }}
            data-testid="search-input"
          />
        </div>
        {null === brands ? (
          <div className="spinner-border text-primary" role="status">
            <span className="sr-only">Loading...</span>
          </div>
        ) : (
          brands.map((brand) => <Brand brand={brand} key={brand.id} />)
        )}

        {page > 1 && (
          <button className="btn btn-dark" onClick={() => {
            dispatch({type: 'PAGE_DOWN'});
          }}>
            PREVIOUS
          </button>
        )}
        <button className="btn btn-dark" onClick={() => {
          dispatch({type: 'PAGE_UP'});
        }}>
          NEXT
        </button>
      </div>
    </BoldContext.Provider>
  );
}

export default App;
