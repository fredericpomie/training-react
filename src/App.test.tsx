import React from "react";
import { render, fireEvent } from "@testing-library/react";
import App from "./App";

test("renders brand list", async () => {
  const { findByText, asFragment, findByTestId } = render(<App />);
  await findByText(/ferrari/i);
  expect(asFragment()).toMatchSnapshot("First loading");

  fireEvent.change(await findByTestId("search-input"), {
    target: { value: "mase" },
  });
  await findByText(/rims maserati/i);
  expect(asFragment()).toMatchSnapshot("after search");

  fireEvent.click(await findByText("NEXT"));
  await findByText(/next/i);
  expect(asFragment()).toMatchSnapshot("after next");
});
