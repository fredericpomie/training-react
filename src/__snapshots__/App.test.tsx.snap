// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders brand list: First loading 1`] = `
<DocumentFragment>
  <div
    class="App container"
  >
    <div
      class="form-group"
    >
      <label>
        Search
      </label>
      <input
        class="form-control"
        data-testid="search-input"
        type="search"
        value=""
      />
    </div>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <img
          alt="Ferrari"
          src="http://img-v2-dev.allopneus.com/brand/label/1.jpg"
        />
      </div>
      <div
        class="col"
      >
        Ferrari
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <img
          alt="Maserati"
          src="http://img-v2-dev.allopneus.com/brand/label/2.gif"
        />
      </div>
      <div
        class="col"
      >
        Maserati
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <img
          alt="Admiral"
          src="http://img-v2-dev.allopneus.com/brand/label/3.jpg"
        />
      </div>
      <div
        class="col"
      >
        Admiral
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <img
          alt="AEZ"
          src="http://img-v2-dev.allopneus.com/brand/label/4.jpg"
        />
      </div>
      <div
        class="col"
      >
        AEZ
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <img
          alt="Alfa Romeo"
          src="http://img-v2-dev.allopneus.com/brand/label/5.jpg"
        />
      </div>
      <div
        class="col"
      >
        Alfa Romeo
      </div>
    </div>
    <button
      class="btn btn-dark"
    >
      NEXT
    </button>
  </div>
</DocumentFragment>
`;

exports[`renders brand list: after next 1`] = `
<DocumentFragment>
  <div
    class="App container"
  >
    <div
      class="form-group"
    >
      <label>
        Search
      </label>
      <input
        class="form-control"
        data-testid="search-input"
        type="search"
        value="mase"
      />
    </div>
    <button
      class="btn btn-dark"
    >
      PREVIOUS
    </button>
    <button
      class="btn btn-dark"
    >
      NEXT
    </button>
  </div>
</DocumentFragment>
`;

exports[`renders brand list: after search 1`] = `
<DocumentFragment>
  <div
    class="App container"
  >
    <div
      class="form-group"
    >
      <label>
        Search
      </label>
      <input
        class="form-control"
        data-testid="search-input"
        type="search"
        value="mase"
      />
    </div>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <img
          alt="Maserati"
          src="http://img-v2-dev.allopneus.com/brand/label/2.gif"
        />
      </div>
      <div
        class="col"
      >
        Maserati
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <img
          alt="Rims Maserati"
          src="http://img-v2-dev.allopneus.com/brand/label/400.jpg"
        />
      </div>
      <div
        class="col"
      >
        Rims Maserati
      </div>
    </div>
    <button
      class="btn btn-dark"
    >
      NEXT
    </button>
  </div>
</DocumentFragment>
`;
